File: hphp/hack/src/hhbc/label_rewriter.ml
(*
 * Copyright (c) 2017, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

open Core_kernel
open Hhbc_ast
open Instruction_sequence

(* Create a map from defined labels to instruction offset *)
let create_label_to_offset_map instrseq =
  snd
  @@ InstrSeq.fold_left instrseq ~init:(0, IMap.empty) ~f:(fun (i, m) instr ->
         match instr with
         | ILabel l -> (i, IMap.add (Label.id l) i m)
         | _ -> (i + 1, m))

let lookup_def l defs =
  match IMap.find_opt l defs with
  | None -> failwith "lookup_def: label missing"
  | Some ix -> ix

(* Get any regular labels referenced by this instruction *)
let get_regular_labels instr =
  match instr with
  | IIterator (IterInit (_, l))
  | IIterator (IterNext (_, l))
  | ICall (FCall (_, _, _, _, Some l, _))
  | ICall (FCallClsMethod ((_, _, _, _, Some l, _), _))
  | ICall (FCallClsMethodD ((_, _, _, _, Some l, _), _, _))
  | ICall (FCallClsMethodS ((_, _, _, _, Some l, _), _))
  | ICall (FCallClsMethodSD ((_, _, _, _, Some l, _), _, _))
  | ICall (FCallFunc (_, _, _, _, Some l, _))
  | ICall (FCallFuncD ((_, _, _, _, Some l, _), _))
  | ICall (FCallObjMethod ((_, _, _, _, Some l, _), _))
  | ICall (FCallObjMethodD ((_, _, _, _, Some l, _), _, _))
  | IGenDelegation (YieldFromDelegate (_, l))
  | IMisc (MemoGet (l, _))
  | IContFlow (Jmp l | JmpNS l | JmpZ l | JmpNZ l) ->
    [l]
  | IContFlow (Switch (_, _, ls)) -> ls
  | IContFlow (SSwitch pairs) -> List.map pairs snd
  | IMisc (MemoGetEager (l1, l2, _)) -> [l1; l2]
  | _ -> []

(* Generate new labels for all labels referenced in instructions and default
 * parameter values, in the same order as used by DumpHhas:
 *   1. First, labels referenced by normal control-flow (jumps, switches, etc)
 *   2. Last, labels referenced by default parameter values
 *)
let create_label_ref_map defs params body =
  let process_ref ((n, (used, refs)) as acc) l =
    let l = Label.id l in
    let ix = lookup_def l defs in
    match IMap.find_opt ix refs with
    (* This is the first time we've seen a reference to a label for
     * this instruction offset, so generate a new label *)
    | None -> (n + 1, (ISet.add l used, IMap.add ix n refs))
    (* We already have a label for this instruction offset *)
    | Some _ -> acc
  in
  let gather_using get_labels acc instrseq =
    InstrSeq.fold_left instrseq ~init:acc ~f:(fun acc instr ->
        List.fold_left (get_labels instr) ~init:acc ~f:process_ref)
  in
  let acc = (0, (ISet.empty, IMap.empty)) in
  let acc = gather_using get_regular_labels acc body in
  let acc =
    List.fold_left params ~init:acc ~f:(fun acc param ->
        match Hhas_param.default_value param with
        | None -> acc
        | Some (l, _) -> process_ref acc l)
  in
  snd acc

let relabel_instr instr relabel =
  match instr with
  | IIterator (IterInit (id, l)) -> IIterator (IterInit (id, relabel l))
  | IIterator (IterNext (id, l)) -> IIterator (IterNext (id, relabel l))
  | IGenDelegation (YieldFromDelegate (i, l)) ->
    IGenDelegation (YieldFromDelegate (i, relabel l))
  | ICall (FCall (fl, na, nr, br, Some l, c)) ->
    ICall (FCall (fl, na, nr, br, Some (relabel l), c))
  | ICall (FCallClsMethod ((fl, na, nr, br, Some l, c), is_log_as_dynamic_call))
    ->
    ICall
      (FCallClsMethod
         ((fl, na, nr, br, Some (relabel l), c), is_log_as_dynamic_call))
  | ICall (FCallClsMethodD ((fl, na, nr, br, Some l, ctx), c, m)) ->
    ICall (FCallClsMethodD ((fl, na, nr, br, Some (relabel l), ctx), c, m))
  | ICall (FCallClsMethodS ((fl, na, nr, br, Some l, ctx), c)) ->
    ICall (FCallClsMethodS ((fl, na, nr, br, Some (relabel l), ctx), c))
  | ICall (FCallClsMethodSD ((fl, na, nr, br, Some l, ctx), c, m)) ->
    ICall (FCallClsMethodSD ((fl, na, nr, br, Some (relabel l), ctx), c, m))
  | ICall (FCallFunc (fl, na, nr, br, Some l, ctx)) ->
    ICall (FCallFunc (fl, na, nr, br, Some (relabel l), ctx))
  | ICall (FCallFuncD ((fl, na, nr, br, Some l, ctx), f)) ->
    ICall (FCallFuncD ((fl, na, nr, br, Some (relabel l), ctx), f))
  | ICall (FCallObjMethod ((fl, na, nr, br, Some l, ctx), f)) ->
    ICall (FCallObjMethod ((fl, na, nr, br, Some (relabel l), ctx), f))
  | ICall (FCallObjMethodD ((fl, na, nr, br, Some l, ctx), f, m)) ->
    ICall (FCallObjMethodD ((fl, na, nr, br, Some (relabel l), ctx), f, m))
  | IContFlow (Jmp l) -> IContFlow (Jmp (relabel l))
  | IContFlow (JmpNS l) -> IContFlow (JmpNS (relabel l))
  | IContFlow (JmpZ l) -> IContFlow (JmpZ (relabel l))
  | IContFlow (JmpNZ l) -> IContFlow (JmpNZ (relabel l))
  | IContFlow (Switch (k, n, ll)) ->
    IContFlow (Switch (k, n, List.map ll relabel))
  | IContFlow (SSwitch pairs) ->
    IContFlow (SSwitch (List.map pairs (fun (id, l) -> (id, relabel l))))
  | IMisc (MemoGet (l, r)) -> IMisc (MemoGet (relabel l, r))
  | IMisc (MemoGetEager (l1, l2, r)) ->
    IMisc (MemoGetEager (relabel l1, relabel l2, r))
  | ILabel l -> ILabel (relabel l)
  | _ -> instr

(* Relabel the instruction sequence and parameter values so that
 *   1. No instruction is preceded by more than one label
 *   2. No label is unreferenced
 *   3. References to labels occur in strict label number order, starting at 0
 *)
let rewrite_params_and_body defs used refs params body =
  let relabel_id l =
    let ix = lookup_def l defs in
    match IMap.find_opt ix refs with
    | None -> failwith "relabel_instrseq: offset not in refs"
    | Some l' -> l'
  in
  (* Rewrite a label that's referenced by an instruction or parameter *)
  let relabel l = Label.map relabel_id l in
  (* Rewrite or remove a label definition *)
  let relabel_define_label_id id =
    if ISet.mem id used then
      IMap.find_opt (lookup_def id defs) refs
    else
      None
  in
  (* Rewrite a single instruction *)
  let rewrite_instr instr =
    match instr with
    | ILabel l ->
      begin
        match Label.option_map relabel_define_label_id l with
        | None -> None
        | Some l' -> Some (ILabel l')
      end
    | _ -> Some (relabel_instr instr relabel)
  in
  (* Rewrite any label referred to in a default value *)
  let rewrite_param param =
    let dv = Hhas_param.default_value param in
    match dv with
    | None -> param
    | Some (l, e) ->
      Hhas_param.make
        (Hhas_param.name param)
        (Hhas_param.is_variadic param)
        (Hhas_param.is_inout param)
        (Hhas_param.user_attributes param)
        (Hhas_param.type_info param)
        (Some (relabel l, e))
  in
  let params = List.map params rewrite_param in
  let body = InstrSeq.filter_map body ~f:rewrite_instr in
  (params, body)

let relabel_function params body =
  let defs = create_label_to_offset_map body in
  let (used, refs) = create_label_ref_map defs params body in
  rewrite_params_and_body defs used refs params body

let clone_with_fresh_regular_labels block =
  let (regular_labels, named_labels) =
    InstrSeq.fold_left
      block
      ~init:(IMap.empty, SMap.empty)
      ~f:(fun ((regular, named) as acc) i ->
        match i with
        | ILabel (Label.Regular _ as l) ->
          (IMap.add (Label.id l) (Label.next_regular ()) regular, named)
        | ILabel (Label.Named name) ->
          (regular, SMap.add name (Label.next_regular ()) named)
        | _ -> acc)
  in
  if IMap.is_empty regular_labels && SMap.is_empty named_labels then
    block
  else
    let relabel l =
      Option.value ~default:l
      @@
      match l with
      | Label.Regular id -> IMap.find_opt id regular_labels
      | Label.Named name -> SMap.find_opt name named_labels
      | _ -> None
    in
    let rewrite_instr instr = relabel_instr instr relabel in
    InstrSeq.map block rewrite_instr

File: hphp/hack/src/hhbc/label_rewriter.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.

use env::emitter::Emitter;
use hhas_param_rust::HhasParam;
use hhbc_ast_rust::{
    FcallArgs, Instruct, InstructCall, InstructControlFlow, InstructIterator, InstructMisc,
};
use instruction_sequence::InstrSeq;
use label_rust::{Id, Label};

use std::collections::{HashMap, HashSet};

fn create_label_to_offset_map(instrseq: &InstrSeq) -> HashMap<Id, usize> {
    let mut folder = |(i, mut map): (usize, HashMap<Id, usize>), instr: &Instruct| match instr {
        Instruct::ILabel(l) => {
            if let Ok(id) = Label::id(l) {
                map.insert(*id, i);
                (i, map)
            } else {
                panic!("Label should've been rewritten by this point")
            }
        }
        _ => (i + 1, map),
    };
    instrseq.fold_left(&mut folder, (0, HashMap::new())).1
}

fn lookup_def<'h>(l: &Id, defs: &'h HashMap<Id, usize>) -> &'h usize {
    match defs.get(l) {
        Some(ix) => ix,
        None => panic!("lookup_def: label missing"),
    }
}

fn get_regular_labels(instr: &Instruct) -> Vec<&Label> {
    use Instruct::*;
    use InstructCall::*;
    use InstructControlFlow::*;
    use InstructIterator::*;
    use InstructMisc::*;
    match instr {
        IIterator(IterInit(_, l))
        | IIterator(IterNext(_, l))
        | IMisc(MemoGet(l, _))
        | IContFlow(Jmp(l))
        | IContFlow(JmpNS(l))
        | IContFlow(JmpZ(l))
        | IContFlow(JmpNZ(l))
        | ICall(FCall(FcallArgs(_, _, _, _, Some(l), _)))
        | ICall(FCallClsMethod(FcallArgs(_, _, _, _, Some(l), _), _))
        | ICall(FCallClsMethodD(FcallArgs(_, _, _, _, Some(l), _), _, _))
        | ICall(FCallClsMethodS(FcallArgs(_, _, _, _, Some(l), _), _))
        | ICall(FCallClsMethodSD(FcallArgs(_, _, _, _, Some(l), _), _, _))
        | ICall(FCallFunc(FcallArgs(_, _, _, _, Some(l), _)))
        | ICall(FCallFuncD(FcallArgs(_, _, _, _, Some(l), _), _))
        | ICall(FCallObjMethod(FcallArgs(_, _, _, _, Some(l), _), _))
        | ICall(FCallObjMethodD(FcallArgs(_, _, _, _, Some(l), _), _, _)) => vec![l],
        IContFlow(Switch(_, _, ls)) => ls.iter().map(|x| x).collect::<Vec<_>>(),
        IContFlow(SSwitch(pairs)) => pairs.iter().map(|x| &x.1).collect::<Vec<_>>(),
        IMisc(MemoGetEager(l1, l2, _)) => vec![l1, l2],
        _ => vec![],
    }
}

fn create_label_ref_map(
    defs: &HashMap<Id, usize>,
    params: &Vec<HhasParam>,
    body: &InstrSeq,
) -> (HashSet<Id>, HashMap<Id, usize>) {
    let process_ref =
        |(mut n, (mut used, mut refs)): (usize, (HashSet<Id>, HashMap<Id, usize>)), l: &Label| {
            if let Ok(id) = Label::id(l) {
                let ix = lookup_def(id, defs);
                if !refs.contains_key(ix) {
                    used.insert(*id);
                    refs.insert(*ix, n);
                    n += 1;
                }
                (n, (used, refs))
            } else {
                panic!("Label should've been rewritten by this point")
            }
        };
    let gather_using = |acc: (usize, (HashSet<Id>, HashMap<Id, usize>)), instrseq: &InstrSeq| {
        let mut folder = |acc: (usize, (HashSet<Id>, HashMap<Id, usize>)), instr: &Instruct| {
            (get_regular_labels(instr))
                .into_iter()
                .fold(acc, process_ref)
        };
        instrseq.fold_left(&mut folder, acc)
    };
    let init = gather_using((0, (HashSet::new(), HashMap::new())), body);
    let (_, map) = params.iter().fold(
        init,
        |acc: (usize, (HashSet<Id>, HashMap<Id, usize>)), param: &HhasParam| match &param
            .default_value
        {
            None => acc,
            Some((l, _)) => process_ref(acc, &l),
        },
    );
    map
}

fn relabel_instr<F>(instr: &mut Instruct, relabel: &mut F)
where
    F: FnMut(&mut Label),
{
    use Instruct::*;
    use InstructCall::*;
    use InstructControlFlow::*;
    use InstructIterator::*;
    use InstructMisc::*;
    match instr {
        IIterator(IterInit(_, l))
        | IIterator(IterNext(_, l))
        | ICall(FCall(FcallArgs(_, _, _, _, Some(l), _)))
        | ICall(FCallClsMethod(FcallArgs(_, _, _, _, Some(l), _), _))
        | ICall(FCallClsMethodD(FcallArgs(_, _, _, _, Some(l), _), _, _))
        | ICall(FCallClsMethodS(FcallArgs(_, _, _, _, Some(l), _), _))
        | ICall(FCallClsMethodSD(FcallArgs(_, _, _, _, Some(l), _), _, _))
        | ICall(FCallFunc(FcallArgs(_, _, _, _, Some(l), _)))
        | ICall(FCallFuncD(FcallArgs(_, _, _, _, Some(l), _), _))
        | ICall(FCallObjMethod(FcallArgs(_, _, _, _, Some(l), _), _))
        | ICall(FCallObjMethodD(FcallArgs(_, _, _, _, Some(l), _), _, _))
        | IContFlow(Jmp(l))
        | IContFlow(JmpNS(l))
        | IContFlow(JmpZ(l))
        | IContFlow(JmpNZ(l))
        | IMisc(MemoGet(l, _))
        | ILabel(l) => relabel(l),
        IContFlow(Switch(_, _, ll)) => ll.iter_mut().for_each(|l| relabel(l)),
        IContFlow(SSwitch(pairs)) => pairs.iter_mut().for_each(|(_, l)| relabel(l)),
        IMisc(MemoGetEager(l1, l2, _)) => {
            relabel(l1);
            relabel(l2);
        }
        _ => {}
    }
}

fn rewrite_params_and_body(
    defs: &HashMap<Id, usize>,
    used: &HashSet<Id>,
    refs: &HashMap<Id, usize>,
    params: &mut Vec<HhasParam>,
    body: &mut InstrSeq,
) {
    let relabel_id = |id: &mut Id| {
        *id = *refs
            .get(lookup_def(&id, defs))
            .expect("relabel_instrseq: offset not in refs")
    };
    let relabel_define_label_id = |id: Id| {
        if used.contains(&id) {
            refs.get(lookup_def(&id, defs)).map(|x| *x)
        } else {
            None
        }
    };
    let mut rewrite_instr = |instr: &mut Instruct| -> bool {
        if let Instruct::ILabel(ref mut l) = instr {
            match l.option_map(relabel_define_label_id) {
                Ok(Some(new_l)) => {
                    *l = new_l;
                    true
                }
                _ => false,
            }
        } else {
            relabel_instr(instr, &mut |l| l.map_mut(relabel_id));
            true
        }
    };
    let rewrite_param = |param: &mut HhasParam| {
        if let Some((l, _)) = &mut param.default_value {
            l.map_mut(relabel_id);
        }
    };
    params.iter_mut().for_each(|param| rewrite_param(param));
    body.filter_map_mut(&mut rewrite_instr);
}

pub fn relabel_function(params: &mut Vec<HhasParam>, body: &mut InstrSeq) {
    let defs = create_label_to_offset_map(body);
    let (used, refs) = create_label_ref_map(&defs, &params, body);
    rewrite_params_and_body(&defs, &used, &refs, params, body)
}

pub fn clone_with_fresh_regular_labels(emitter: &mut Emitter, block: &mut InstrSeq) {
    let mut folder = |
        (mut regular, mut named): (HashMap<Id, Label>, HashMap<String, Label>),
        instr: &Instruct,
    | {
        match instr {
            Instruct::ILabel(Label::Regular(id)) => {
                regular.insert(*id, emitter.label_gen_mut().next_regular());
            }
            Instruct::ILabel(Label::Named(name)) => {
                named.insert(name.to_string(), emitter.label_gen_mut().next_regular());
            }
            _ => {}
        }
        (regular, named)
    };
    let (regular_labels, named_labels) =
        block.fold_left(&mut folder, (HashMap::new(), HashMap::new()));

    if !regular_labels.is_empty() || !named_labels.is_empty() {
        let relabel = |l: &mut Label| {
            let new_label = match l {
                Label::Regular(id) => regular_labels.get(id),
                Label::Named(name) => named_labels.get(name),
                _ => None,
            };
            if let Some(nl) = new_label {
                *l = nl.clone();
            }
        };
        block.map_mut(&mut |instr| relabel_instr(instr, &mut |l| relabel(l)))
    }
}
